File: arena-junit/src/main/java/arena/junit/playbook/ActiveHttpPlaybook.java
package arena.junit.playbook;

import arena.junit.ffi.ArenaBindings;
import arena.junit.ffi.ArenaBindingError;
import arena.junit.support.ArenaJson;

import com.fasterxml.jackson.databind.node.ObjectNode;
import com.sun.jna.Pointer;

public class ActiveHttpPlaybook extends ActivePlaybook {
  protected ActiveHttpPlaybook(Pointer handle) {
    super(handle);
  }

  public void verify(String method, String urlPath, int expectedCount) {
    ObjectNode spec = ArenaJson.object();
    spec.put("method", method.toUpperCase());
    spec.put("url_path", urlPath);
    spec.put("expected_count", expectedCount);
    try {
      ArenaBindings.httpPlaybookVerify(handle(), ArenaJson.MAPPER.writeValueAsString(spec));
    } catch (ArenaBindingError e) {
      noteBodyFailure();
      throw e;
    } catch (Exception e) {
      throw new ArenaBindingError(e.getMessage());
    }
  }

  public void verifyAtLeast(String method, String urlPath, int minimumCount) {
    ObjectNode spec = ArenaJson.object();
    spec.put("method", method.toUpperCase());
    spec.put("url_path", urlPath);
    spec.put("minimum_count", minimumCount);
    try {
      ArenaBindings.httpPlaybookVerify(handle(), ArenaJson.MAPPER.writeValueAsString(spec));
    } catch (ArenaBindingError e) {
      noteBodyFailure();
      throw e;
    } catch (Exception e) {
      throw new ArenaBindingError(e.getMessage());
    }
  }
}

File: arena-xunit/src/Playbook/ActiveHttpPlaybook.cs
using System;
using ArenaDotnet.Xunit.Ffi;
using ArenaDotnet.Xunit.Support;
using Newtonsoft.Json;

namespace ArenaDotnet.Xunit.Playbook;

public sealed class ActiveHttpPlaybook : ActivePlaybook
{
    public ActiveHttpPlaybook(IntPtr handle) : base(handle) { }

    public void Verify(string method, string path, int expectedCount)
    {
        var spec = ArenaJson.Serialize(new VerifySpec
        {
            Method = method,
            Path = path,
            ExpectedCount = expectedCount,
        });
        ArenaBindings.HttpPlaybookVerify(_handle, spec);
    }

    public void VerifyAtLeast(string method, string path, int minCount)
    {
        var spec = ArenaJson.Serialize(new VerifyAtLeastSpec
        {
            Method = method,
            Path = path,
            MinCount = minCount,
        });
        ArenaBindings.HttpPlaybookVerify(_handle, spec);
    }

    [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
    private sealed class VerifySpec
    {
        [JsonProperty("method")] public string? Method { get; set; }
        [JsonProperty("url_path")] public string? Path { get; set; }
        [JsonProperty("expected_count")] public int? ExpectedCount { get; set; }
    }

    [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
    private sealed class VerifyAtLeastSpec
    {
        [JsonProperty("method")] public string? Method { get; set; }
        [JsonProperty("url_path")] public string? Path { get; set; }
        [JsonProperty("minimum_count")] public int? MinCount { get; set; }
    }
}
